pkg/kubecli/vmi.go

```diff
@@ -245,8 +246,59 @@ func (ws *wsStreamer) Stream(options StreamOptions) error {
 func (v *vmis) VNC(name string) (StreamInterface, error) {
 	return v.asyncSubresourceHelper(name, "vnc")
 }
-func (v *vmis) SerialConsole(name string) (StreamInterface, error) {
-	return v.asyncSubresourceHelper(name, "console")
+
+type connectionStruct struct {
+	con StreamInterface
+	err error
+}
+
+func (v *vmis) SerialConsole(name string, timeout time.Duration) (StreamInterface, error) {
+	timeoutChan := time.Tick(timeout)
+	connectionChan := make(chan connectionStruct)
+	isWaiting := true
+
+	go func() {
+		con, err := v.asyncSubresourceHelper(name, "console")
+		for err != nil && isWaiting {
+			if asyncSubresourceError, ok := err.(*AsyncSubresourceError); ok {
+				if asyncSubresourceError.GetStatusCode() == http.StatusBadRequest {
+					// Sleep to prevent denial of service on the api server
+					time.Sleep(1 * time.Second)
```

Comment:
I think it is better to use watches instead of polling. That should be much cheaper once the connection is established.